{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import os\n",
    "import glob"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def get_entity_dict(a1_file):\n",
    "    entity_dict={}\n",
    "    with open(a1_file, 'r') as reader:\n",
    "        for line in reader:\n",
    "            tab_del_list=line.strip().split(\"\\t\")\n",
    "            entity_id = tab_del_list[0]\n",
    "            entity_type = tab_del_list[1].split(' ')[0]\n",
    "            start_off = int(tab_del_list[1].split(' ')[1])\n",
    "            end_off = int(tab_del_list[1].split(' ')[2])\n",
    "            entity_name = tab_del_list[2]\n",
    "            \n",
    "            #print entity_id, entity_type,start_off, end_off,entity_name\n",
    "            if entity_id not in entity_dict:\n",
    "                entity_dict[entity_id] = [entity_type, start_off, end_off,entity_name]\n",
    "            elif entity_id in entity_dict:\n",
    "                raise Exception('key error')\n",
    "                \n",
    "    return entity_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def get_event_dict(a2_file):\n",
    "    event_dict={}\n",
    "    with open(a2_file, 'r') as reader:\n",
    "        for line in reader:\n",
    "            if 'E' in line.strip().split('\\t')[0]:\n",
    "                event_id = line.strip().split('\\t')[0]\n",
    "                if event_id not in event_dict:\n",
    "                    event_dict[event_id]=line.strip().split('\\t')[1].split(' ')\n",
    "                else:\n",
    "                    raise Exception('key error')  \n",
    "\n",
    "                \n",
    "    return event_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def extract_entity_relation(line, event_dict, entity_dict):\n",
    "    if 'E' in line.split('\\t')[0]:\n",
    "        #print line.split('\\t')[0],line.split('\\t')[1].split(' ')\n",
    "        if len(line.split('\\t')[1].split(' '))>=3:\n",
    "            theme_string = line.split('\\t')[1].split(' ')[1].split(':')[1]\n",
    "            cause_string = line.split('\\t')[1].split(' ')[2].split(':')[1]\n",
    "            \n",
    "            #print line\n",
    "            if 'E' in theme_string:\n",
    "                effect = event_dict[theme_string][0].split(':')[0]\n",
    "                result_entity_id = event_dict[theme_string][1].split(':')[1]\n",
    "\n",
    "                if 'E' in cause_string:\n",
    "                    cause_entity_id = event_dict[cause_string][1].split(':')[1]\n",
    "                    if 'T' in cause_entity_id and 'T' in result_entity_id:\n",
    "                        return entity_dict[cause_entity_id][3] , effect ,  entity_dict[result_entity_id][3]\n",
    "                \n",
    "                elif 'E' not in cause_string :\n",
    "                    cause_entity_id = line.split('\\t')[1].split(' ')[2].split(':')[1]\n",
    "                    if 'T' in result_entity_id:\n",
    "                        return entity_dict[cause_entity_id][3] , effect , entity_dict[result_entity_id][3]\n",
    "\n",
    "            elif 'E' in cause_string:\n",
    "                    effect = line.split('\\t')[1].split(' ')[0].split(':')[0]\n",
    "                    result_entity_id = theme_string\n",
    "                    cause_entity_id = event_dict[cause_string][1].split(':')[1]\n",
    "                    \n",
    "                    if 'T' in cause_entity_id: \n",
    "                        return entity_dict[cause_entity_id][3] , effect , entity_dict[result_entity_id][3]\n",
    "\n",
    "\n",
    "def preprocess(a1_file, a2_file):\n",
    "    entity_dict = get_entity_dict(a1_file)\n",
    "    event_dict = get_event_dict(a2_file)\n",
    "    \n",
    "    result_list =[]\n",
    "    \n",
    "    with open(a2_file, 'r') as reader:\n",
    "        for line in reader:\n",
    "            line = line.strip()\n",
    "            result = extract_entity_relation(line,event_dict,entity_dict)\n",
    "            if result is not None:\n",
    "                result_list.append(result)\n",
    "                \n",
    "    return result_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "text = './data/BioNLP-ST-2013_GE_train_data_rev3/PMC-1310901-01-INTRODUCTION.txt'\n",
    "a1 = './data/BioNLP-ST-2013_GE_train_data_rev3/PMC-1310901-01-INTRODUCTION.a1'\n",
    "a2 = './data/BioNLP-ST-2013_GE_train_data_rev3/PMC-1310901-01-INTRODUCTION.a2'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "preprocess(a1, a2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "dataset_folder = './data/BioNLP-ST-2013_GE_train_data_rev3/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 171,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "file_list = list(set(map(lambda x:x.split('.')[0],os.listdir(dataset_folder))))\n",
    "file_list.remove('LICENSE')\n",
    "file_list.remove('README')\n",
    "path_list = map(lambda x:dataset_folder+x , file_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 226,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "total_list=[]\n",
    "for path in path_list:\n",
    "    a1_file = path+'.a1'\n",
    "    a2_file = path+'.a2'\n",
    "    total_list.extend(preprocess(a1_file, a2_file))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 229,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('IL-2', 'Gene_expression', 'IL-2'),\n",
       " ('IL-2', 'Localization', 'IL-2'),\n",
       " ('Bcl10', 'Protein_catabolism', 'Bcl10'),\n",
       " ('PKCalpha', 'Phosphorylation', 'p65'),\n",
       " ('Foxp3', 'Binding', 'CREB-1'),\n",
       " ('Foxp3', 'Binding', 'activating transcription factor 2'),\n",
       " ('p300', 'Transcription', 'Gal4-BD-CREB-1'),\n",
       " ('Foxp3', 'Positive_regulation', 'Gal4-BD-CREB-1'),\n",
       " ('Foxp3', 'Positive_regulation', 'Gal4-BD-CREB-1'),\n",
       " ('p300', 'Binding', 'p300'),\n",
       " ('Foxp3', 'Binding', 'p300'),\n",
       " ('Foxp3', 'Binding', 'p300'),\n",
       " ('LMP1', 'Positive_regulation', 'Akt'),\n",
       " ('Akt', 'Negative_regulation', 'FoxO1'),\n",
       " ('STAT4', 'Gene_expression', 'IL-10'),\n",
       " ('STAT4', 'Gene_expression', 'IFN-gamma'),\n",
       " ('STAT4', 'Gene_expression', 'IL-4'),\n",
       " ('STAT4', 'Gene_expression', 'IL-10'),\n",
       " ('STAT4', 'Gene_expression', 'IL-10'),\n",
       " ('IL-12', 'Gene_expression', 'IL-10'),\n",
       " ('OVA', 'Gene_expression', 'IL-10'),\n",
       " ('STAT6', 'Gene_expression', 'IL-4'),\n",
       " ('STAT6', 'Gene_expression', 'IL-10'),\n",
       " ('STAT4', 'Gene_expression', 'IL-10'),\n",
       " ('IFN-gamma', 'Regulation', 'IL-12'),\n",
       " ('IFN-gamma', 'Gene_expression', 'IL-10'),\n",
       " ('IFN-gamma', 'Localization', 'IL-4'),\n",
       " ('IL-4', 'Gene_expression', 'IL-10'),\n",
       " ('STAT4', 'Gene_expression', 'IL-10'),\n",
       " ('STAT6', 'Gene_expression', 'IL-10'),\n",
       " ('ERK1', 'Positive_regulation', 'p38'),\n",
       " ('ERK2', 'Positive_regulation', 'p38'),\n",
       " ('ERK1', 'Positive_regulation', 'GSK3beta'),\n",
       " ('ERK2', 'Positive_regulation', 'GSK3beta'),\n",
       " ('IL-12', 'Positive_regulation', 'ERK1'),\n",
       " ('IL-12', 'Positive_regulation', 'ERK2'),\n",
       " ('ERK1', 'Positive_regulation', 'IL-10'),\n",
       " ('ERK2', 'Positive_regulation', 'IL-10'),\n",
       " ('MEK5', 'Positive_regulation', 'ERK5'),\n",
       " ('MEK1', 'Positive_regulation', 'ERK1'),\n",
       " ('MEK2', 'Positive_regulation', 'ERK1'),\n",
       " ('MEK1', 'Positive_regulation', 'ERK2'),\n",
       " ('MEK2', 'Positive_regulation', 'ERK2'),\n",
       " ('p38 MAPK', 'Gene_expression', 'IL-10'),\n",
       " ('GSK3beta', 'Gene_expression', 'IL-10'),\n",
       " ('IL-12', 'Gene_expression', 'IL-10'),\n",
       " ('ERK1', 'Gene_expression', 'IL-10'),\n",
       " ('ERK2', 'Gene_expression', 'IL-10'),\n",
       " ('STAT-4', 'Gene_expression', 'IL-10'),\n",
       " ('IL-12', 'Positive_regulation', 'STAT-4'),\n",
       " ('ERK1', 'Gene_expression', 'IL-10'),\n",
       " ('ERK2', 'Gene_expression', 'IL-10'),\n",
       " ('IL-10', 'Gene_expression', 'IL-10'),\n",
       " ('STAT4', 'Positive_regulation', 'IL-10'),\n",
       " ('IL-12', 'Positive_regulation', 'IL-10'),\n",
       " ('IL-12', 'Gene_expression', 'IL-10'),\n",
       " ('IFN-gamma', 'Positive_regulation', 'IL-10'),\n",
       " ('IL-12', 'Gene_expression', 'IL-10'),\n",
       " ('IFN-gamma', 'Positive_regulation', 'IL-10'),\n",
       " ('IL-12', 'Positive_regulation', 'IL-10'),\n",
       " ('ERK1', 'Gene_expression', 'IL-10'),\n",
       " ('ERK2', 'Gene_expression', 'IL-10'),\n",
       " ('IL-12', 'Gene_expression', 'Gata-3'),\n",
       " ('IL-12', 'Gene_expression', 'c-maf'),\n",
       " ('ERK', 'Gene_expression', 'c-maf'),\n",
       " ('ERK', 'Positive_regulation', 'IL-10'),\n",
       " ('c-Maf', 'Gene_expression', 'IL-10'),\n",
       " ('ERK1', 'Gene_expression', 'IL-10'),\n",
       " ('ERK2', 'Gene_expression', 'IL-10'),\n",
       " ('IL-12', 'Gene_expression', 'IL-10'),\n",
       " ('IL-12', 'Gene_expression', 'IL-10'),\n",
       " ('IKKbeta', 'Phosphorylation', 'p65'),\n",
       " ('IKKbeta', 'Phosphorylation', 'p65'),\n",
       " ('PKCalpha', 'Transcription', 'IL-1alpha'),\n",
       " ('PKCalpha', 'Transcription', 'iNOS'),\n",
       " ('PKCalpha', 'Transcription', 'TNFalpha'),\n",
       " ('LMP1', 'Positive_regulation', 'IL15'),\n",
       " ('LMP1', 'Positive_regulation', 'IFNgamma'),\n",
       " ('IL10', 'Positive_regulation', 'Stat3'),\n",
       " ('IL10', 'Positive_regulation', 'Stat3'),\n",
       " ('LMP1', 'Positive_regulation', 'IL10'),\n",
       " ('LMP1', 'Positive_regulation', 'IL10'),\n",
       " ('IL10', 'Positive_regulation', 'pStat3'),\n",
       " ('pStat3', 'Localization', 'pStat3'),\n",
       " ('Pam3CYS', 'Gene_expression', 'IFNgamma'),\n",
       " ('Sharpin', 'Negative_regulation', 'Sharpin'),\n",
       " ('IL-12', 'Gene_expression', 'IL-10'),\n",
       " ('IL-12', 'Gene_expression', 'IL-10'),\n",
       " ('IL-12', 'Positive_regulation', 'IL-10'),\n",
       " ('IL-12p40', 'Positive_regulation', 'IL-10'),\n",
       " ('IL-12', 'Gene_expression', 'IL-10'),\n",
       " ('IL-12', 'Gene_expression', 'IL-10'),\n",
       " ('OVA', 'Gene_expression', 'IL-10'),\n",
       " ('IL-12', 'Gene_expression', 'IL-10'),\n",
       " ('IL-12', 'Gene_expression', 'IL-10'),\n",
       " ('PKCalpha', 'Phosphorylation', 'p65'),\n",
       " ('PKCalpha', 'Phosphorylation', 'p65'),\n",
       " ('PKCalpha', 'Phosphorylation', 'p65'),\n",
       " ('PKCalpha', 'Phosphorylation', 'p65'),\n",
       " ('PKCalpha siRNA', 'Gene_expression', 'PKCalpha'),\n",
       " ('Foxp3', 'Localization', 'Foxp3'),\n",
       " ('Foxp3', 'Binding', 'Foxp3'),\n",
       " ('Sharpin', 'Ubiquitination', 'RIP1'),\n",
       " ('Sharpin', 'Ubiquitination', 'NEMO'),\n",
       " ('HOIP', 'Positive_regulation', 'IKKgamma'),\n",
       " ('IKKgamma', 'Negative_regulation', 'IKKgamma'),\n",
       " ('IKKgamma', 'Negative_regulation', 'IKKgamma'),\n",
       " ('IKKgamma', 'Positive_regulation', 'IKKgamma'),\n",
       " ('IKKgamma', 'Positive_regulation', 'IKKgamma'),\n",
       " ('HOIP', 'Binding', 'CD40'),\n",
       " ('HOIP', 'Binding', 'CD40'),\n",
       " ('TRAF2', 'Binding', 'HOIP'),\n",
       " ('TRAF2', 'Binding', 'HOIP'),\n",
       " ('SHARPIN', 'Transcription', 'IL6'),\n",
       " ('SHARPIN', 'Transcription', 'GMCSF'),\n",
       " ('SHARPIN', 'Gene_expression', 'IL6'),\n",
       " ('SHARPIN', 'Gene_expression', 'GMCSF'),\n",
       " ('IL4', 'Phosphorylation', 'Stat6'),\n",
       " ('ERK1', 'Gene_expression', 'IL-10'),\n",
       " ('ERK2', 'Gene_expression', 'IL-10'),\n",
       " ('ERK1', 'Gene_expression', 'IL-10'),\n",
       " ('ERK2', 'Gene_expression', 'IL-10'),\n",
       " ('ERK1', 'Gene_expression', 'IL-10'),\n",
       " ('ERK2', 'Gene_expression', 'IL-10'),\n",
       " ('MEK', 'Gene_expression', 'IL-10'),\n",
       " ('OVA', 'Gene_expression', 'IL-10'),\n",
       " ('Foxp3', 'Binding', 'Foxp3'),\n",
       " ('Foxp3', 'Localization', 'Foxp3'),\n",
       " ('Foxp3', 'Binding', 'Foxp3'),\n",
       " ('Foxp3', 'Negative_regulation', 'Foxp3'),\n",
       " ('Tax', 'Gene_expression', 'Foxp3'),\n",
       " ('Foxp3', 'Positive_regulation', 'IL-2'),\n",
       " ('T-bet', 'Gene_expression', 'interferon (IFN) gamma'),\n",
       " ('Eomesodermin', 'Gene_expression', 'interferon (IFN) gamma'),\n",
       " ('Runx3', 'Gene_expression', 'Eomes'),\n",
       " ('Runx3', 'Gene_expression', 'IFN-gamma'),\n",
       " ('Runx3', 'Gene_expression', 'perforin'),\n",
       " ('Runx3', 'Gene_expression', 'granzyme B'),\n",
       " ('LMP1', 'Gene_expression', 'IL10'),\n",
       " ('STAT4', 'Gene_expression', 'Interleukin-10'),\n",
       " ('ERK', 'Gene_expression', 'Interleukin-10'),\n",
       " ('IL-12', 'Positive_regulation', 'STAT4'),\n",
       " ('ERK1', 'Gene_expression', 'IL-10'),\n",
       " ('ERK2', 'Gene_expression', 'IL-10'),\n",
       " ('Runx3', 'Gene_expression', 'Eomes'),\n",
       " ('Runx3', 'Gene_expression', 'perforin'),\n",
       " ('Runx3', 'Gene_expression', 'granzyme B'),\n",
       " ('Runx3', 'Gene_expression', 'IFN-gamma'),\n",
       " ('T-bet', 'Gene_expression', 'IFN-gamma'),\n",
       " ('Runx3', 'Gene_expression', 'Eomes'),\n",
       " ('Runx3', 'Gene_expression', 'perforin'),\n",
       " ('Runx3', 'Gene_expression', 'granzyme B'),\n",
       " ('Runx3', 'Gene_expression', 'IFN-gamma'),\n",
       " ('Eomes', 'Gene_expression', 'IFN-gamma'),\n",
       " ('Runx3', 'Gene_expression', 'perforin'),\n",
       " ('Eomes', 'Gene_expression', 'perforin'),\n",
       " ('T-bet', 'Gene_expression', 'perforin'),\n",
       " ('T-bet', 'Gene_expression', 'Gzmb'),\n",
       " ('T-bet', 'Gene_expression', 'granzyme B'),\n",
       " ('dominant negative PKCalpha', 'Phosphorylation', 'p65'),\n",
       " ('dominant negative PKCalpha', 'Phosphorylation', 'p65'),\n",
       " ('TNFalpha', 'Phosphorylation', 'p65'),\n",
       " ('PKCalpha', 'Phosphorylation', 'p65'),\n",
       " ('ubiquitin', 'Protein_catabolism', 'p65'),\n",
       " ('p65', 'Binding', 'p65'),\n",
       " ('IKKalpha', 'Gene_expression', 'p52'),\n",
       " ('p65', 'Positive_regulation', 'p65'),\n",
       " ('LMP1', 'Gene_expression', 'IL10'),\n",
       " ('IL10', 'Positive_regulation', 'Stat3'),\n",
       " ('LMP1', 'Positive_regulation', 'Akt'),\n",
       " ('LMP1', 'Positive_regulation', 'cRel'),\n",
       " ('CD40', 'Negative_regulation', 'p27'),\n",
       " ('pAkt', 'Phosphorylation', 'GSK3'),\n",
       " ('pAkt', 'Negative_regulation', 'GSK3'),\n",
       " ('Akt', 'Negative_regulation', 'Fox01'),\n",
       " ('FoxO1', 'Gene_expression', 'bcl6'),\n",
       " ('LMP1', 'Gene_expression', 'LMP1'),\n",
       " ('bcl6', 'Gene_expression', 'LMP1'),\n",
       " ('LMP1', 'Gene_expression', 'bcl6'),\n",
       " ('bcl6', 'Gene_expression', 'bcl6'),\n",
       " ('Fox01', 'Regulation', 'bcl6'),\n",
       " ('Runx3', 'Positive_regulation', 'Eomes'),\n",
       " ('Runx3', 'Positive_regulation', 'granzyme B'),\n",
       " ('Runx3', 'Positive_regulation', 'perforin'),\n",
       " ('Runx3', 'Positive_regulation', 'IFN-gamma'),\n",
       " ('Runx3', 'Gene_expression', 'granzyme B'),\n",
       " ('Runx3', 'Gene_expression', 'perforin'),\n",
       " ('Runx3', 'Gene_expression', 'IFN-gamma'),\n",
       " ('Runx3', 'Gene_expression', 'TNF'),\n",
       " ('Runx3', 'Gene_expression', 'IL-2'),\n",
       " ('Runx3', 'Gene_expression', 'IFN-gamma'),\n",
       " ('Runx1', 'Positive_regulation', 'TNF'),\n",
       " ('Runx1', 'Positive_regulation', 'IL-2'),\n",
       " ('Runx3', 'Gene_expression', 'IFN-gamma'),\n",
       " ('T-bet', 'Gene_expression', 'IFN-gamma'),\n",
       " ('Eomes', 'Gene_expression', 'IFN-gamma'),\n",
       " ('T-bet', 'Gene_expression', 'IFN-gamma'),\n",
       " ('T-bet', 'Gene_expression', 'perforin'),\n",
       " ('Eomes', 'Gene_expression', 'perforin'),\n",
       " ('T-bet', 'Gene_expression', 'granzyme B'),\n",
       " ('Eomes', 'Gene_expression', 'perforin'),\n",
       " ('T-bet', 'Positive_regulation', 'Gzmb'),\n",
       " ('Eomes', 'Positive_regulation', 'Prf1'),\n",
       " ('T-bet', 'Gene_expression', 'perforin'),\n",
       " ('Eo-VP16', 'Gene_expression', 'granzyme B'),\n",
       " ('Eomes', 'Gene_expression', 'perforin'),\n",
       " ('Runx3', 'Gene_expression', 'perforin'),\n",
       " ('Runx3', 'Positive_regulation', 'Eomes'),\n",
       " ('T-bet', 'Regulation', 'Ifng'),\n",
       " ('Eomes', 'Regulation', 'Ifng'),\n",
       " ('T-bet', 'Regulation', 'Prf1'),\n",
       " ('Eomes', 'Regulation', 'Prf1'),\n",
       " ('T-bet', 'Regulation', 'GzmB'),\n",
       " ('Eomes', 'Regulation', 'GzmB'),\n",
       " ('T-bet', 'Gene_expression', 'IFN-gamma'),\n",
       " ('Eomes', 'Gene_expression', 'perforin'),\n",
       " ('Foxp3', 'Binding', 'p300'),\n",
       " ('Foxp3', 'Binding', 'Foxp3'),\n",
       " ('Foxp3', 'Binding', 'Foxp3'),\n",
       " ('Tax', 'Binding', 'Tax'),\n",
       " ('Tax', 'Gene_expression', 'Foxp3'),\n",
       " ('Foxp3', 'Binding', 'CBP'),\n",
       " ('Foxp3', 'Binding', 'p300'),\n",
       " ('CD40', 'Phosphorylation', 'IkappaBalpha'),\n",
       " ('CD40', 'Protein_catabolism', 'IkappaBalpha'),\n",
       " ('type I IFN',\n",
       "  'Positive_regulation',\n",
       "  'signal transducer and activator of transcription (STAT)1'),\n",
       " ('IFN-gamma',\n",
       "  'Positive_regulation',\n",
       "  'signal transducer and activator of transcription (STAT)1'),\n",
       " ('signal transducer and activator of transcription (STAT)1',\n",
       "  'Gene_expression',\n",
       "  'IFN-gamma'),\n",
       " ('STAT3', 'Gene_expression', 'RORgammat'),\n",
       " ('IL-12', 'Gene_expression', 'IFN-gamma'),\n",
       " ('ERK1', 'Localization', 'IL-4'),\n",
       " ('ERK2', 'Localization', 'IL-4'),\n",
       " ('IL-10', 'Gene_expression', 'IL-10'),\n",
       " ('ERK1', 'Gene_expression', 'IL-10'),\n",
       " ('ERK2', 'Gene_expression', 'IL-10'),\n",
       " ('Eomes', 'Gene_expression', 'IFN-gamma'),\n",
       " ('Runx3', 'Gene_expression', 'Eomes'),\n",
       " ('Runx3', 'Gene_expression', 'perforin'),\n",
       " ('T-bet', 'Gene_expression', 'perforin'),\n",
       " ('Runx3', 'Gene_expression', 'IFN-gamma'),\n",
       " ('Runx3', 'Gene_expression', 'granzyme B'),\n",
       " ('Runx3', 'Gene_expression', 'Gzmb'),\n",
       " ('T-bet', 'Transcription', 'Ifng'),\n",
       " ('Eomes', 'Transcription', 'Prf1'),\n",
       " ('HOIP', 'Binding', 'IKKgamma'),\n",
       " ('CD40', 'Positive_regulation', 'CD80'),\n",
       " ('CD40', 'Gene_expression', 'CD80'),\n",
       " ('CD40', 'Positive_regulation', 'CD80'),\n",
       " ('CD40', 'Positive_regulation', 'CD80'),\n",
       " ('HOIP', 'Positive_regulation', 'CD80'),\n",
       " ('HOIP', 'Positive_regulation', 'CD80'),\n",
       " ('T-bet', 'Gene_expression', 'Perforin'),\n",
       " ('T-bet', 'Gene_expression', 'granzyme B'),\n",
       " ('T-bet', 'Transcription', 'perforin'),\n",
       " ('T-bet', 'Negative_regulation', 'GzmB'),\n",
       " ('Eo-VP16', 'Gene_expression', 'perforin'),\n",
       " ('T-bet', 'Gene_expression', 'IFN-gamma'),\n",
       " ('Eo-VP16', 'Transcription', 'GzmB'),\n",
       " ('T-bet', 'Transcription', 'GzmB'),\n",
       " ('T-bet', 'Gene_expression', 'perforin'),\n",
       " ('Eomes', 'Gene_expression', 'perforin'),\n",
       " ('T-bet', 'Gene_expression', 'granzyme B'),\n",
       " ('Eomes', 'Gene_expression', 'granzyme B'),\n",
       " ('T-bet', 'Gene_expression', 'IFN-gamma'),\n",
       " ('Eomes', 'Gene_expression', 'IFN-gamma'),\n",
       " ('Runx3', 'Gene_expression', 'IFN-gamma'),\n",
       " ('Runx3', 'Gene_expression', 'TNF'),\n",
       " ('Runx3', 'Gene_expression', 'IL-2'),\n",
       " ('Runx3', 'Positive_regulation', 'Eomes'),\n",
       " ('Runx3', 'Positive_regulation', 'Eomes'),\n",
       " ('Runx3', 'Transcription', 'perforin'),\n",
       " ('Eomes', 'Transcription', 'perforin'),\n",
       " ('Runx3', 'Transcription', 'IFN-gamma'),\n",
       " ('Eomes', 'Transcription', 'IFN-gamma'),\n",
       " ('Runx3', 'Transcription', 'granzyme B'),\n",
       " ('Eomes', 'Transcription', 'granzyme B'),\n",
       " ('Runx3', 'Gene_expression', 'Eomes'),\n",
       " ('Runx3', 'Gene_expression', 'perforin'),\n",
       " ('Runx3', 'Gene_expression', 'granzyme B'),\n",
       " ('Runx3', 'Gene_expression', 'IFN-gamma'),\n",
       " ('Eo-VP16', 'Gene_expression', 'perforin'),\n",
       " ('Runx3', 'Gene_expression', 'perforin'),\n",
       " ('Eomes', 'Gene_expression', 'perforin'),\n",
       " ('Runx3', 'Gene_expression', 'perforin'),\n",
       " ('Eomes', 'Gene_expression', 'perforin'),\n",
       " ('Runx3', 'Gene_expression', 'granzyme B'),\n",
       " ('Eomes', 'Gene_expression', 'granzyme B'),\n",
       " ('IL-12', 'Gene_expression', 'IFN-gamma'),\n",
       " ('IL-12', 'Gene_expression', 'IL-10'),\n",
       " ('IL-12', 'Gene_expression', 'IFN-gamma'),\n",
       " ('IL-12', 'Gene_expression', 'IL-10'),\n",
       " ('IL-12', 'Gene_expression', 'IL-10'),\n",
       " ('IL-12', 'Gene_expression', 'IL-10'),\n",
       " ('IL-12', 'Gene_expression', 'IL-10'),\n",
       " ('TGF-beta', 'Gene_expression', 'IL-10'),\n",
       " ('IL-12', 'Gene_expression', 'IL-10'),\n",
       " ('IL-2', 'Gene_expression', 'IL-10'),\n",
       " ('TGF-beta', 'Gene_expression', 'IL-10'),\n",
       " ('IL-12', 'Transcription', 'Ifngamma'),\n",
       " ('IL-12', 'Transcription', 'Ifngamma'),\n",
       " ('IL-12', 'Transcription', 'Il10'),\n",
       " ('IL-12', 'Gene_expression', 'Ifngamma'),\n",
       " ('IL-12', 'Transcription', 'GATA-3'),\n",
       " ('IL-12', 'Transcription', 'c-Maf'),\n",
       " ('IL-12', 'Gene_expression', 'c-maf'),\n",
       " ('c-maf', 'Positive_regulation', 'STAT4'),\n",
       " ('MEK1', 'Gene_expression', 'Il10'),\n",
       " ('MEK2', 'Gene_expression', 'Il10'),\n",
       " ('MEK1', 'Gene_expression', 'c-maf'),\n",
       " ('MEK2', 'Gene_expression', 'c-maf'),\n",
       " ('MEK1', 'Gene_expression', 'T-bet'),\n",
       " ('MEK2', 'Gene_expression', 'T-bet'),\n",
       " ('MEK1', 'Gene_expression', 'RORgammat'),\n",
       " ('MEK2', 'Gene_expression', 'RORgammat'),\n",
       " ('Foxp3', 'Localization', 'Foxp3'),\n",
       " ('Foxp3', 'Gene_expression', 'p65'),\n",
       " ('deltaFKH', 'Gene_expression', 'p65'),\n",
       " ('IFN-alpha', 'Gene_expression', 'IRF-4'),\n",
       " ('IRF-4', 'Positive_regulation', 'IRF-4'),\n",
       " ('c-Rel', 'Gene_expression', 'IRF-4'),\n",
       " ('IRF-4', 'Binding', 'IRF-4'),\n",
       " ('TRAF2', 'Binding', 'HOIL-1L-interacting protein'),\n",
       " ('CD40', 'Positive_regulation', 'CD80'),\n",
       " ('Runx3', 'Gene_expression', 'CD4'),\n",
       " ('Foxp3', 'Gene_expression', 'Tax'),\n",
       " ('deltaFKH', 'Positive_regulation', 'CREB'),\n",
       " ('TRAF2', 'Binding', 'HOIL-1L interacting protein'),\n",
       " ('CD40', 'Positive_regulation', 'CD80'),\n",
       " ('Foxp3', 'Gene_expression', 'Tax'),\n",
       " ('Foxp3', 'Transcription', 'Tax'),\n",
       " ('deltaFKH', 'Gene_expression', 'Tax'),\n",
       " ('Foxp3', 'Gene_expression', 'Tax'),\n",
       " ('HOIP', 'Binding', 'CD40'),\n",
       " ('TRAF2', 'Binding', 'HOIP'),\n",
       " ('cIAP1', 'Binding', 'HOIP'),\n",
       " ('cIAP2', 'Binding', 'HOIP'),\n",
       " ('cIAP1', 'Binding', 'HOIP'),\n",
       " ('cIAP2', 'Binding', 'HOIP'),\n",
       " ('cIAP', 'Binding', 'HOIP'),\n",
       " ('cIAP', 'Protein_modification', 'HOIP'),\n",
       " ('HOIP', 'Binding', 'CD40'),\n",
       " ('HOIP', 'Binding', 'TRAF2'),\n",
       " ('HOIP', 'Binding', 'TRAF3'),\n",
       " ('HOIP', 'Binding', 'cIAP1'),\n",
       " ('HOIP', 'Binding', 'IKKgamma'),\n",
       " ('HOIP', 'Binding', 'IKKgamma'),\n",
       " ('IKKgamma', 'Protein_modification', 'IKKgamma')]"
      ]
     },
     "execution_count": 229,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "total_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import nltk.data\n",
    "import re\n",
    "tokenizer = nltk.data.load('tokenizers/punkt/english.pickle')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def get_sentence_list(text_file):\n",
    "    sentence_list = []\n",
    "    with open(text_file,'r') as reader:\n",
    "        for i,line in enumerate(reader):\n",
    "            temp_list = tokenizer.tokenize(line.strip())\n",
    "            sentence_list.extend(temp_list)  \n",
    "    return sentence_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def get_entity_set(a1_file):\n",
    "    with open(a1_file ,'r') as reader:\n",
    "        entity_set = set([line.strip().split('\\t')[2] for line in reader])\n",
    "    return entity_set"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def get_word_positions(string, word):\n",
    "    position_list=[]\n",
    "    for m in re.finditer(word, string):\n",
    "        position_list.append((m.start(),m.end()))\n",
    "    return position_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def insert_entitiy_notation(text_file,a1_file):\n",
    "    entity_dict = get_entity_dict(a1_file)\n",
    "    entity_set = get_entity_set(a1_file)\n",
    "    sentence_list = get_sentence_list(text_file)\n",
    "    \n",
    "    for sentence in sentence_list:\n",
    "        for entity in entity_set:\n",
    "            position_list = get_word_positions(sentence,entity)\n",
    "            if len(position_list)>=1:\n",
    "                print sentence, entity, position_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The hallmark genetic aberration of CML is a reciprocal chromosomal translocation t(9;22) leading to expression of a bcr-abl fusion gene, an aberrant activated tyrosine kinase (2). bcr-abl fusion gene [(116, 135)]\n",
      "The hallmark genetic aberration of CML is a reciprocal chromosomal translocation t(9;22) leading to expression of a bcr-abl fusion gene, an aberrant activated tyrosine kinase (2). bcr [(116, 119)]\n",
      "Treatment with interferon alpha (IFN-alpha) prolongs survival of CML patients and is associated with a complete cytogenetic response in 5-33% of CML patients (1,2). interferon alpha [(15, 31)]\n",
      "Treatment with interferon alpha (IFN-alpha) prolongs survival of CML patients and is associated with a complete cytogenetic response in 5-33% of CML patients (1,2). IFN-alpha [(33, 42)]\n",
      "Recently, we described an impaired expression of the interferon regulatory factor 4 (IRF-4) in CML, correlating with poor response to IFN-alpha treatment (3). IFN-alpha [(134, 143)]\n",
      "Recently, we described an impaired expression of the interferon regulatory factor 4 (IRF-4) in CML, correlating with poor response to IFN-alpha treatment (3). interferon regulatory factor 4 [(53, 83)]\n",
      "Recently, we described an impaired expression of the interferon regulatory factor 4 (IRF-4) in CML, correlating with poor response to IFN-alpha treatment (3). IRF-4 [(85, 90)]\n",
      "The cause of the silencing of IRF-4 level remained unclear. IRF-4 [(30, 35)]\n",
      "IRF-4 (ICSAT/Pip/MUM1/LSIRF) is one member with very restricted expression pattern: Predominately B- and activated T-lymphocytes are IRF-4 positive (7-11). ICSAT [(7, 12)]\n",
      "IRF-4 (ICSAT/Pip/MUM1/LSIRF) is one member with very restricted expression pattern: Predominately B- and activated T-lymphocytes are IRF-4 positive (7-11). MUM1 [(17, 21)]\n",
      "IRF-4 (ICSAT/Pip/MUM1/LSIRF) is one member with very restricted expression pattern: Predominately B- and activated T-lymphocytes are IRF-4 positive (7-11). LSIRF [(22, 27)]\n",
      "IRF-4 (ICSAT/Pip/MUM1/LSIRF) is one member with very restricted expression pattern: Predominately B- and activated T-lymphocytes are IRF-4 positive (7-11). Pip [(13, 16)]\n",
      "IRF-4 (ICSAT/Pip/MUM1/LSIRF) is one member with very restricted expression pattern: Predominately B- and activated T-lymphocytes are IRF-4 positive (7-11). IRF-4 [(0, 5), (133, 138)]\n",
      "In contrast to other IRFs, expression of IRF-4 cannot be induced by IFNs, but by antigen stimulation, crosslinking of T- or B-cell receptors or phorbol-myristate-acetate (10,11). IRF-4 [(41, 46)]\n",
      "Consistent with the restriction of expression to immunocompetent cells, mice with deletion of IRF-4 failed to develop mature and functionally active B- and T-lymphocytes (12), and the impaired expression of IRF-4 in CML was predominately found in T-cells (3). IRF-4 [(94, 99), (207, 212)]\n",
      "These data suggest a crucial role for IRF-4 in the function of immune cells. IRF-4 [(38, 43)]\n",
      "Examples for such aberrated genes are MGMT, DAPK, p14ARF, p15INK4b, p16INK4a, BRCA1, CDH13 and APAF-1 (17-19). BRCA1 [(78, 83)]\n",
      "Examples for such aberrated genes are MGMT, DAPK, p14ARF, p15INK4b, p16INK4a, BRCA1, CDH13 and APAF-1 (17-19). p16INK4a [(68, 76)]\n",
      "Examples for such aberrated genes are MGMT, DAPK, p14ARF, p15INK4b, p16INK4a, BRCA1, CDH13 and APAF-1 (17-19). DAPK [(44, 48)]\n",
      "Examples for such aberrated genes are MGMT, DAPK, p14ARF, p15INK4b, p16INK4a, BRCA1, CDH13 and APAF-1 (17-19). CDH13 [(85, 90)]\n",
      "Examples for such aberrated genes are MGMT, DAPK, p14ARF, p15INK4b, p16INK4a, BRCA1, CDH13 and APAF-1 (17-19). p15INK4b [(58, 66)]\n",
      "Examples for such aberrated genes are MGMT, DAPK, p14ARF, p15INK4b, p16INK4a, BRCA1, CDH13 and APAF-1 (17-19). APAF-1 [(95, 101)]\n",
      "Examples for such aberrated genes are MGMT, DAPK, p14ARF, p15INK4b, p16INK4a, BRCA1, CDH13 and APAF-1 (17-19). MGMT [(38, 42)]\n",
      "Examples for such aberrated genes are MGMT, DAPK, p14ARF, p15INK4b, p16INK4a, BRCA1, CDH13 and APAF-1 (17-19). p14ARF [(50, 56)]\n",
      "In CML, methylation is known to regulate expression of the c-abl, the bcr gene and others (20-23), and the extent of methylation in the c-abl promoter has been shown to be associated with advanced disease (24). c-abl [(59, 64), (136, 141)]\n",
      "In CML, methylation is known to regulate expression of the c-abl, the bcr gene and others (20-23), and the extent of methylation in the c-abl promoter has been shown to be associated with advanced disease (24). bcr [(70, 73)]\n",
      "In this work, we studied mechanisms of IRF-4 gene expression silencing in leukemic cells. IRF-4 [(39, 44)]\n",
      "We analyzed the IRF-4 promoter region for genetic aberrations and methylational status in IRF-4-positive and -negative hematopoietic cells. IRF-4 [(16, 21), (90, 95)]\n",
      "Wall time: 3 ms\n"
     ]
    }
   ],
   "source": [
    "%time insert_entitiy_notation(text, a1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
